src/crimson/os/seastore/onode_manager/staged-fltree/node.h

```diff
@@ -0,0 +1,341 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#pragma once
+
+#include <map>
+#include <memory>
+#include <ostream>
+#include <boost/smart_ptr/intrusive_ref_counter.hpp>
+
+#include "crimson/common/type_helpers.h"
+
+#include "node_extent_mutable.h"
+#include "stages/key_layout.h"
+#include "stages/stage_types.h"
+#include "super.h"
+#include "tree_types.h"
+
+namespace crimson::os::seastore::onode {
+
+/**
+ * in-memory subtree management:
+ *
+ * resource management (bottom-up):
+ * USER          --> Ref<tree_cursor_t>
+ * tree_cursor_t --> Ref<LeafNode>
+ * Node (child)  --> Ref<InternalNode> (see parent_info_t)
+ * Node (root)   --> Super::URef
+ * Super         --> Btree
+ *
+ * tracked lookup (top-down):
+ * Btree         --> Super*
+ * Super         --> Node* (root)
+ * InternalNode  --> Node* (children)
+ * LeafNode      --> tree_cursor_t*
+ */
+
+class LeafNode;
+class InternalNode;
+
```

Comment:
I think I mostly follow what's going on here, but I think every public method could use a formal interface comment explaining the behavior.  In particular, lookup_* have specific behaviour when empty that should be spelled out.

Reply:
fixed in https://github.com/ceph/ceph/pull/36986/commits/1c053f3425ac9bc2c3d88d2ffcd6d85d18aa7f9f